### PR TITLE
add order

### PR DESCRIPTION
**Pull request by ptiforka**
controller
model
spec

### CODE REVIEW

File: spec/controllers/api/orders_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+        it 'user is not winner' do
+          subject
+          expect(json_response[:errors][:user].first).to eq('you dont winner at this lot')
+        end
+      end
+    end
+  end
+
+  describe '#sent' do
```
Comment (VladRezonans) [inline spec/controllers/api/orders_controller_spec.rb]: create two describe

File: app/models/order.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  belongs_to :user
+  belongs_to :lot
+  belongs_to :bid, optional: true, dependent: :destroy
+  before_create :check_status, :check_user
+  after_create :set_order_id, :send_email
+  after_destroy :new_winner_lot
+
+  private
+
+  def new_winner_lot
```
Comment (VladRezonans) [inline app/models/order.rb]: remove

File: app/models/order.rb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    all_bids.each do |bid|
+      if bid.proposed_price > lot.current_price
+        lot.current_price = bid.proposed_price
+        lot.winning_bid = bid.id
+      end
+    end
+    lot.closed!
+  end
+
+  def set_order_id
```
Comment (VladRezonans) [inline app/models/order.rb]: remove

File: app/models/order.rb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      errors.add :order, "you cant create order at status #{lot.status}"
+    end
+  end
+
+  def check_user
+    winner_user_id = Bid.find_by_id(lot.winning_bid).user_id
+    errors.add :user, 'you dont winner at this lot' if user_id != winner_user_id
+  end
+
+  def send_email
```
[the remark this answers is not in the document]
Comment (ptiforka) [reply app/models/order.rb]: +

File: app/models/order.rb [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    OrderMailer.send_confirmation(self).deliver
+  end
+
+  def send_status_email
+    OrderMailer.send_status(self).deliver
+  end
+
+  public
```
[the remark this answers is not in the document]
Comment (ptiforka) [reply app/models/order.rb]: +

File: app/models/order.rb [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    if pending? && current_user.id == lot.user_id
+      sent!
+      send_status_email
+    else
+      errors.add :order, 'status or user wrong'
+    end
+  end
+
+  def deliver!(current_user)
+    if sent? && current_user.id == Bid.find_by_id(lot.winning_bid).user_id
```
[the remark this answers is not in the document]
Comment (ptiforka) [reply app/models/order.rb]: +

File: app/policies/order_policy.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  end
+
+  def destroy?
+    true if @user.present? && @user.id == bid_user_id && @order.pending?
+  end
+
+  private
+
+  def bid_user_id
+    @bid = Bid.find(@order.bid_id).user_id
```
[the remark this answers is not in the document]
Comment (ptiforka) [reply app/policies/order_policy.rb]: +

File: app/views/order_mailer/send_confirmation.html.erb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+<h1>Order: #<%= @order.id %></h1>
+
+<p>You ordered  lot:</p>
+
+<ul>
+  title:"
+  <%=@order.lot.title %>
+  price:
+  <%=@order.lot.current_price %>
+</ul>
```
[the remark this answers is not in the document]
Comment (ptiforka) [reply app/views/order_mailer/send_confirmation.html.erb]: +

File: app/views/order_mailer/send_status.html.erb [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+<ul>
+  title:
+  <%=@order.lot.title %>
+  price:
+  <%=@order.lot.current_price %>
+
+  order status changed to
+  <%= @order.status %>
+</ul>
```
[the remark this answers is not in the document]
Comment (ptiforka) [reply app/views/order_mailer/send_status.html.erb]: +

File: app/controllers/v1/orders_controller.rb [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+class V1::OrdersController < ApplicationController
+  before_action :authenticate_user!
+
+  def create
+    order = current_user.orders.create(order_params)
+    render_resource(order)
+  end
+
+  def update
+    authorize order
```
[the remark this answers is not in the document]
Comment (ptiforka) [reply app/controllers/v1/orders_controller.rb]: +

File: app/models/lot.rb [hunk 11]
```diff
@@ -8,11 +8,14 @@ class Lot < ApplicationRecord
...
-  after_create_commit :add_job
+  after_create :add_job
 
   belongs_to :user
+  has_one :order
   has_many :bids, dependent: :destroy
 
+  private
+
   def add_job
```
[the remark this answers is not in the document]
Comment (ptiforka) [reply app/models/lot.rb]: +

File: spec/controllers/api/orders_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+    end
+
+    before(:each) do
+      lot.winning_bid = bid.id
+      lot.closed!
+    end
+
+    context 'user creator lot' do
+      it 'should change status to sent' do
+        lot.winning_bid = bid.id
```
[the remark this answers is not in the document]
Comment (ptiforka) [reply spec/controllers/api/orders_controller_spec.rb]: +

File: app/models/order.rb (hunk 3 above)
Comment (ptiforka) [reply app/models/order.rb]: +

File: app/models/order.rb (hunk 2 above)
Comment (ptiforka) [reply app/models/order.rb]: +

File: spec/controllers/api/orders_controller_spec.rb (hunk 1 above)
Comment (ptiforka) [reply spec/controllers/api/orders_controller_spec.rb]: +

File: spec/controllers/api/orders_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,188 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe V1::OrdersController, type: :request do
+  describe '#create' do
+    let(:bid) { FactoryBot.create :bid }
+    let(:lot) { Lot.find(bid.lot_id) }
```
[the remark this answers is not in the document]
Comment (ptiforka) [reply spec/controllers/api/orders_controller_spec.rb]: +

**Pull request merged**